### ISSUE TITLE
Create a Group from a user list

### ISSUE THREAD

**Issue by hwatheod** [learning-unlimited/ESP-Website#2007]
Add an interface for creating a new Group of users from an arbitrary user list, either a list generated by the "Arbitrary User List" functionality, or just a copy/pasted list of usernames.  This is useful when known groups of students (for example, for Stanford, outreach students) need special permissions like extended deadlines, etc.

(Right now, this can only be done either one-by-one per student through the admin interface, or needs someone with server access to run a script to add a whole list.)

**Comment by milescalabresi** [comment]
Clark has made this same request. I wonder if it wouldn't be cleaner to cut out the groups entirely and just create permissions for a user list. Such a model would avoid clutter in the list of groups (which are more fundamental) and would allow an easier interface when adding new permissions.

**Comment by willgearty** [comment]
Doesn't that have the potential to create an enormous number of permissions, which could be hard to look through? Whereas assigning a new user group only adds a single group and single permissions for whichever permissions are wanted? Since user groups are listed on the userview page, that would probably be an easier way to figure out what permissions a user has?

**Comment by benjaminjkraft** [comment]
I don't love either approach and I think this is going to need some more design thought.  In particular, having groups feels really ugly for me because groups are designed to be a more permanent entity -- they aren't dated in any way -- so allowing them to proliferate isn't great.  On the other hand, in the existing system there's no good way to describe a user list in plain English, which makes it hard to put on the permissions page collectively, and adding individual permissions to that page (or excluding these permissions) seems like a nightmare.  One way to avoid the proliferation of groups would be to have long-lived groups, and just add the users to one of those, and grant it permissions.  But that seems like it's bound to cause problems if people reuse accounts or anything.

So stepping back a bit: what do you think we *should* display on the permissions page if we've granted permissions to a collection of users?  Is there anything we want to be able to do with these collections of users other than grant them permissions?  If one wants to add a user to one of these collections, how should one do it?

**Comment by willgearty** [comment]
I could envision a ManyToMany situation where you can apply a permission to a whole batch of users as just a single permission entry?
Then a single entry would list all of the users that that permission is now applied to?
And since it's a ManyToMany, you could even have multiple permissions applied to a single user (or batch of users) as a single entry, right?
This should also make it fairly easy to add a user to a permission, maybe? Although it could be more complicated if this was to be a UI thing.

At Stanford at least, we also use these user groups for emailing and printing (schedules and nametags), so I'd be hesitant to totally get rid of groups at this point, unless there's a reasonable replacement/alternative.

**Comment by milescalabresi** [comment]
I like that idea, but it seems difficult to modify in real time, for example, if a single user changes status but is part of a group that is granted a single "batch permission." Would the admin have to go back and recreate the batch permission for all other users in that group because a single person changes status?

**Comment by willgearty** [comment]
Could you give an example of when you would need to change the permissions for a single user because of a status change?
Most permissions that we enforce at Stanford are for a specific program, and they expire after the program, so you don't need to worry about users changing status.

However, even if you do, I suppose you could remove them from the batch permission through the admin panel? Then create a unique set of permissions for that user?

**Comment by milescalabresi** [comment]
For example, suppose you have a more limited set of lunch options for students with a certain dietary restriction, so the site can automatically keep them from signing up for foods containing harmful ingredient (say, peanuts). If a student emails you and says they incorrectly reported their dietary information, you'd want to move them back to a different group. Since this kind of information is open-ended by its nature, you'd need to verify the group manually.

As for a separate unique set, that's true, but since the messiness of which permissions override others is already confusing, I like to put in a vote for future changes to permissions improving the clarity if not the UI for how permissions interact with one another and with deadlines.

**Comment by willgearty** [comment]
Resurrecting this discussion because of a post on slack. @milescalabresi, wouldn't the groups setup be better for the example you are suggesting, since you can just add or remove users to/from a group on an individual basis. Implementing what @hwatheod suggests would just add a way to batch add users to a group for when the groups are initiated (and it could even be flexible enough to add users to an existing group or add only the users that aren't already in the group).

**Comment by milescalabresi** [comment]
It depends if under your scheme permissions update automatically every time the group is updated. But I also agree with Ben's point that groups should not proliferate or else other things could get messy

**Comment by willgearty** [comment]
The permission would be associated with the group, so when you change who is in the group, the permissions get applied to the new people in the group and not the people removed from the group. Isn't the whole point of Django groups to combine people who should all have the same permissions? So isn't this what they are designed for?

**Comment by jacklat** [comment]
I don't have many firm ideas as far as implementation goes, but I did just want to comment on this to say that it's a feature that I'd also really like to see/which has recently come up for our program (certain 7th grade classes are full, so it would be helpful to be able to alter permissions so as to close registration for 7th graders and it seems as if there is currently no great way to do so). 

From a user's point of view (obviously there are probably other concerns as far as development/actual implementation goes), I don't think it would make a particular difference to me whether the option to subselect user lists goes under groups vs permissions (or even under the arbitrary user lists themselves--i.e. if after selecting a list, there was a button to alter a permission for that list rather than just view it). However, having the list as dynamic rather than static is definitely important to me (i.e. if we create a block on all future 7th graders registering for a program, that block should apply to some new 7th grader who creates a profile on the site rather than just all 7th graders currently existing in the database).

**Comment by willgearty** [comment]
@jacklat That functionality is already implemented. See here: https://github.com/learning-unlimited/ESP-Website/blob/main/docs/admin/releases/08/README.rst#per-grade-program-caps

**Comment by milescalabresi** [comment]
my understanding was the permissions from groups were meant to be more permanent/intrinsic, not something that people would frequently join and leave the way we do deadlines, but I think it would work

**Comment by willgearty** [comment]
Hi from SplashCon!

OK, so I think I'm seeing two broad ideas here:

1. Make it easier to generate and add users to groups (ideally with a module that uses the UserSearchController)
2. Make permissions accept some sort of user filter (e.g. all 7th graders) that would always be up-to-date (if a new 7th grade student creates an account, the permission should apply to them).

The former should be pretty easy (and we could have an option to remove users from the group that are no longer selected in the user list). The latter will probably be a bit harder since that's not really how permissions work at the moment.
